crates/sui-framework-build/src/compiled_package.rs

```diff
@@ -474,3 +492,76 @@ impl PackageHooks for SuiPackageHooks {
         Ok(())
     }
 }
+
+enum AddressError {
+    NoPublishedAddress(Symbol),
+    InvalidPublishedAddress(Symbol, String),
+}
+
+impl fmt::Display for AddressError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            AddressError::NoPublishedAddress(name) => write!(
+                f,
+                "Package dependency \"{name}\" does not \
+		     specify a published address \
+		     (the Move.toml manifest for \"{name}\" does \
+		     not contain a published-at field)."
+            ),
+            AddressError::InvalidPublishedAddress(name, value) => write!(
+                f,
+                "Package dependency \"{name}\" does not \
+		     specify a valid published address: \
+		     could not parse value \"{value}\" for published-at field.",
+            ),
+        }
+    }
+}
+
+/// Ensure dependencies are published (contain `published-at` address in manifest).
+pub fn ensure_published_dependencies(resolution_graph: &ResolvedGraph) -> Result<(), SuiError> {
+    let mut address_errors = vec![];
+    for name in resolution_graph.graph.package_table.keys() {
+        if let Some(package) = &resolution_graph.package_table.get(name) {
+            let value = package
+                .source_package
+                .package
+                .custom_properties
+                .get(&Symbol::from(PUBLISHED_AT_MANIFEST_FIELD));
+
+            if value.is_none() {
+                address_errors.push(AddressError::NoPublishedAddress(*name));
+                continue;
+            }
+
+            if value
+                .and_then(|v| ObjectID::from_str(v.as_str()).ok())
+                .is_none()
+            {
+                address_errors.push(AddressError::InvalidPublishedAddress(
+                    *name,
+                    value.unwrap().to_string(),
+                ));
+                continue;
+            }
+        }
+    }
+    if !address_errors.is_empty() {
+        let mut error_messages = address_errors
+            .iter()
+            .map(|v| format!("{}", v))
+            .collect::<Vec<_>>();
+        error_messages.push(
+            "If this is intentional, you may use the \
+	     --with-unpublished-dependencies flag to \
+	     continue publishing these packages as part \
+	     of your package (they won't be linked to \
+	     existing packages on-chain)."
```

Comment:
Nit: this is a particularly tightly wrapped string now, could we spread it out a bit?  Also some small wording changes, but otherwise this looks great!

```suggestion
            "If this is intentional, you may use the \
	     --with-unpublished-dependencies flag to \
	     continue publishing these dependencies as part \
	     of your package (they won't be linked against \
	     existing packages on-chain)."
```